react/components/Popper.tsx

```diff
@@ -0,0 +1,160 @@
+  useCallback,
+  MutableRefObject,
+} from 'react'
+import {
+  State,
+  Instance,
+  createPopper,
+  VirtualElement,
+  ModifierArguments,
+  Placement as PopperPlacementType,
+} from '@popperjs/core'
+
+import Portal from './Portal'
+import setRef from '../modules/setRef'
+import useForkRef from '../modules/useForkRef'
+
+interface Props {
+  anchorEl?:
+    | null
+    | VirtualElement
+    | (() => VirtualElement)
+    | React.RefObject<HTMLElement>
+  children: React.ReactNode
+  open?: boolean
+  placement?: PopperPlacementType
+  transition?: boolean
+  role?: string
+}
+
+function getAnchorEl(anchorEl: React.ReactNode) {
+  return typeof anchorEl === 'function' ? anchorEl() : anchorEl
+}
+
+const Popper = React.forwardRef(function Popper(props: Props, ref) {
+  const {
+    role,
+    anchorEl,
+    children,
+    open = false,
+    transition = false,
+    placement: initialPlacement = 'bottom',
+  } = props
+
+  const tooltipRef = useRef<HTMLElement>(null)
+  const ownRef = useForkRef(tooltipRef, ref)
+  const popperRef: MutableRefObject<Instance | null> = useRef<Instance>(null)
+  const [placement, setPlacement] = useState(initialPlacement)
+  const [exited, setExited] = useState(true)
+
+  // It doesn't have a dependency array because it should update
+  // after every render of react
+  useEffect(() => {
+    if (popperRef.current) {
+      popperRef.current.update()
+    }
+  })
+
+  const handleOpen = useCallback(() => {
+    if (!tooltipRef.current || !open) {
+      return
+    }
+
+    if (popperRef.current) {
+      popperRef.current.destroy()
+    }
+
+    const handlePopperFirstUpdate = (data: Partial<State>) => {
+      if (data.placement) {
+        setPlacement(data.placement)
+      }
+    }
+
+    const handlePopperUpdate = (data: ModifierArguments<{}>) => {
+      setPlacement(data.state.placement)
+    }
+
+    const popper = createPopper(getAnchorEl(anchorEl), tooltipRef.current, {
+      placement: initialPlacement,
+      onFirstUpdate: handlePopperFirstUpdate,
+      modifiers: [
+        {
+          name: 'afterWrite',
+          enabled: true,
+          phase: 'afterWrite',
+          fn: handlePopperUpdate,
+        },
+      ],
+    })
+    popperRef.current = popper
+  }, [open, initialPlacement, anchorEl])
+
+  const handleRef = useCallback(
+    node => {
+      setRef(ownRef, node)
+      handleOpen()
+    },
+    [ownRef, handleOpen]
+  )
+
+  const handleEnter = () => {
+    setExited(false)
+  }
+
+  const handleClose = () => {
+    if (!popperRef.current) {
+      return
+    }
+
+    popperRef.current.destroy()
+  }
+
+  const handleExited = () => {
+    setExited(true)
+    handleClose()
+  }
+
+  useEffect(() => {
+    handleOpen()
+  }, [handleOpen])
+
+  useEffect(() => {
+    return () => {
+      handleClose()
+    }
+  }, [])
+
+  useEffect(() => {
+    if (!open && !transition) {
+      handleClose()
+    }
+  }, [open, transition])
+
+  if (!open && (!transition || exited)) {
```

Comment:
This `exited` variable might not make any sense now but it will follow the same logic of the [BaseModal](https://github.com/vtex-apps/modal-layout/blob/master/react/BaseModal.tsx#L128) component